### PR TITLE
[Routing] Allow non string route requirements

### PR DESCRIPTION
**Pull request by dawehner**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #12430 |
| License | MIT |
| Doc PR |  |

### CODE REVIEW

File: src/Symfony/Component/Routing/Route.php [hunk 1]
```diff
@@ -567,7 +582,7 @@ public function compile()
 
     private function sanitizeRequirement($key, $regex)
     {
-        if (!is_string($regex)) {
+        if (!is_string($regex) && $this->enforceRequirementAsString) {
```
Comment (iltar) [inline src/Symfony/Component/Routing/Route.php]: Is it still a regex if it's not a string? If not, the local parameter should be renamed imo

Comment (dawehner) [inline src/Symfony/Component/Routing/Route.php]: You are absolute right!

File: src/Symfony/Component/Routing/Route.php [hunk 2]
```diff
@@ -565,24 +580,24 @@ public function compile()
...
+    private function sanitizeRequirement($key, $regex_or_mixed)
     {
-        if (!is_string($regex)) {
+        if (!is_string($regex_or_mixed) && $this->enforceRequirementAsString) {
             throw new \InvalidArgumentException(sprintf('Routing requirement for "%s" must be a string.', $key));
         }
 
-        if ('' !== $regex && '^' === $regex[0]) {
-            $regex = (string) substr($regex, 1); // returns false for a single character
+        if ('' !== $regex_or_mixed && '^' === $regex_or_mixed[0]) {
```
Comment (Tobion) [inline src/Symfony/Component/Routing/Route.php]: this will fail if it is not a string.

File: src/Symfony/Component/Routing/Route.php [hunk 3]
```diff
@@ -565,24 +580,24 @@ public function compile()
         return $this->compiled = $class::compile($this);
     }
 
-    private function sanitizeRequirement($key, $regex)
+    private function sanitizeRequirement($key, $regex_or_mixed)
```
Comment (Tobion) [inline src/Symfony/Component/Routing/Route.php]: we use camelCase

**Comment by Tobion** [comment]
The phpdoc and variable name of setRequirement would need to be changed. And the RouteCompiler as well to handle non-string requirements.

File: src/Symfony/Component/Routing/Route.php [hunk 4]
```diff
@@ -65,6 +70,11 @@ class Route implements \Serializable
     private $condition = '';
 
     /**
+     * @var bool
+     */
+    private $enforceRequirementAsString = TRUE;
```
Comment (hhamon) [inline src/Symfony/Component/Routing/Route.php]: should be `true`

**Comment by dawehner** [comment]
Thank you @weaverryan for the good conversation! The latest changes adapt to that

**Comment by xabbuh** [comment]
@dawehner Looks like some tests need to be satisfied first.

Status: Needs work

**Comment by dawehner** [comment]
Tests are fixed now, sadly the generated router doesn't comply with the CS

File: src/Symfony/Component/Routing/Tests/RouteTest.php [hunk 5]
```diff
@@ -135,13 +135,32 @@ public function getInvalidRequirements()
...
+        $route = new Route('/{foo}');
+        $route->setRequirement('foo', $req);
+        $this->assertEquals($req, $route->getRequirement('foo'));
+    }
+
+    public function getValidRequirements()
+    {
+        return array(
+            array(TRUE),
+            array(FALSE),
```
Comment (Tobion) [inline src/Symfony/Component/Routing/Tests/RouteTest.php]: lowercase please

**Comment by fabpot** [comment]
@dawehner no need to worry about the CS for generated code, we don't really care.

File: src/Symfony/Component/Routing/Tests/RouteTest.php (hunk 5 above)
Comment (dawehner) [inline src/Symfony/Component/Routing/Tests/RouteTest.php]: fixed it

**Comment by nicolas-grekas** [comment]
Isn't this problematic? With this patch, people will add requirements that are not expected by the `getRequirements` consumers, which will likely fail with some hard to understand error message.
I'd prefer making `sanitizeRequirements` protected personally...

**Comment by Tobion** [comment]
> I'd prefer making sanitizeRequirements protected personally...

That would not solve the problem as a subclass would still violate the contract if it allows non-strings at the end.

**Comment by nicolas-grekas** [comment]
Yep, but then it's up to subclasser to use their child class in a way that makes sense.
The current approach pushes the burden on the base code, which looks worse to me.

**Comment by stof** [comment]
I see an issue here: storing a non-string value in a requirement as done by Drupal means using the requirement for something totally different than their current meaning (i.e. a regex used to match the routing placeholder).
Btw, a clear indication of this is that ``_access`` is not a routing placeholder in their case.

IMO, they should be storing this in route options or route defaults rather than in route requirements.

**Comment by stof** [comment]
so I'm voting -1 on this PR.

**Comment by dawehner** [comment]
What about making the method protected instead?

**Comment by dawehner** [comment]
Much like symfony we also cannot just change our routing access API completly

**Comment by dawehner** [comment]
Is there any way to move this issue forward in general?

**Comment by nicolas-grekas** [comment]
Closing as :-1: 
Alternative way proposed in  #19077

**Comment by nicolas-grekas** [comment]
For reference: If one adds a non string requirement and use it in the pattern, things explode. If one dumps routes (see command in FrameworkBundle), things explode. In fact, all existing code expecting a string will explode.

**Comment by dawehner** [comment]
At some point this PR had an opt IN behaviour. Could we start with that again? This should not break BC, while, as discussed in two threads now, is sadly less nice.

**Pull request closed without merging**